File: 2020/15.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    start = [16, 11, 15, 0, 1, 7]
    spoken = Dict{Int, Int}()
    for (i, v) in enumerate(start)
        spoken[v] = i
    end
    last = start[end]
    return spoken, last
end

function play(spoken::Dict{Int, Int}, last_spoken::Int, turns::Int)
    turn = length(spoken)
    while turn < turns
        if !haskey(spoken, last_spoken)
            spoken[last_spoken] = turn
            last_spoken = 0
        else
            spoken[last_spoken], last_spoken = turn, turn - spoken[last_spoken]
        end
        turn += 1
    end
    last_spoken
end

function part1(spoken::Dict{Int, Int}, last::Int)
    println("🎄 Part 1: $(play(spoken, last, 2020))")
end

function part2(spoken::Dict{Int, Int}, last::Int)
    println("🎄🎅 Part 2: $(play(spoken, last, 30000000))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 15        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

spoken, last = loadInput()
@time part1(spoken, last)
spoken, last = loadInput()
@time part2(spoken, last)

File: 2023/15.py
import time
from utils import readInput

def loadInput():
    with open("input_15.txt", "r") as f:
        line = f.readline()
    instructions = line[:-1].split(",")
    return instructions

def hash_algo(instr):
    val = 0
    for c in instr:
        val += ord(c)
        val *= 17
        val %= 256
    return val
    
def part1(inputs):
    codes = []
    for instr in inputs:
        codes.append(hash_algo(instr))
    return sum(codes)

def focusing_power(boxes):
    val = 0
    for nb, box in boxes.items():
        val += sum([(nb+1) * (i+1) * b for i,b in enumerate(box.values())])
    return val
    
def part2(inputs):
    boxes = {k:{} for k in range(0, 256)}
    for instr in inputs:
        add = True
        if "=" in instr:
            label, value = instr.split("=")
        else:
            add = False
            label = instr[:-1]
        box = hash_algo(label)
        if not add and label in boxes[box]:
            del boxes[box][label]
        elif add:
            boxes[box][label] = int(value)
                
    return focusing_power(boxes)

if __name__ == '__main__':
    print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
    print("⛄ Day 15: Lens Library ⛄")
    print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
    
    inputs = loadInput()

    t0 = time.time()
    res1 = part1(inputs)
    t1 = time.time()-t0
    
    t0 = time.time()
    res2 = part2(inputs)
    t2 = time.time()-t0
    
    print (f"🎄 Part 1: {res1} ({t1:.5f}) - 🎄🎅 Part 2: {res2} ({t2:.5f})")
